### PR TITLE
Update to Chaquopy 14, remove old stdio redirection mechanism

### PR DESCRIPTION
**Pull request by mhsmith**
DO NOT MERGE until Chaquopy 14.0.0 has been released publicly on Maven Central, and the temporary Maven URL in build.gradle has been removed.

Things to check:
* The native stdout and stderr redirection to logcat has now been incorporated into Chaquopy. This can be tested using something like `ctypes.CDLL(ctypes.util.find_library("c")).printf(...)`.
* The app build no longer requires CMake or the NDK. If they aren't already present in the Android SDK, they should not be installed.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
These changes all make sense, and the improved build time (and lack of CMake warnings) is awesome. One comment inline about a possible tweak. 

The only other thing I noticed was that stdout still has some linebreaks - for example:
```
01-20 08:58:05.037 17133 17133 D MainActivity: Running main module tests.helloworld
01-20 08:58:05.154 17133 17133 I python.stdout: ============================= test session starts ==============================
01-20 08:58:05.154 17133 17133 I python.stdout: platform linux -- Python 3.10.6, pytest-7.2.1, pluggy-1.0.0 -- /system/bin/app_process64
01-20 08:58:05.154 17133 17133 I python.stdout: cachedir: /data/user/0/com.example.helloworld/cache/chaquopy/tmp/.pytest_cache
01-20 08:58:05.154 17133 17133 I python.stdout: rootdir: /data/data/com.example.helloworld/files/chaquopy/AssetFinder/app
01-20 08:58:05.154 17133 17133 I python.stdout: collecting ... 
01-20 08:58:05.155 17133 17133 I python.stdout: collected 1 item
01-20 08:58:05.156 17133 17133 I python.stdout: 
01-20 08:58:05.156 17133 17133 I python.stdout: tests/test_app.py::test_first 
01-20 08:58:05.156 17133 17133 I python.stdout: PASSED                                     [100%]
01-20 08:58:05.157 17133 17133 I python.stdout: 
01-20 08:58:05.157 17133 17133 I python.stdout: 
01-20 08:58:05.157 17133 17133 I python.stdout: ============================== 1 passed in 0.00s ===============================
```
or, on a longer test suite:
```
01-20 09:37:08.600  3201  3245 I python.stdout: 
01-20 09:37:08.600  3201  3245 I python.stdout: tests/widgets/test_slider.py::test_change 
01-20 09:37:08.601  3201  3245 I python.stdout: SKIPPED (on_change called 0 times) [ 84%]
01-20 09:37:08.601  3201  3245 I python.stdout: 
01-20 09:37:08.601  3201  3245 I python.stdout: tests/widgets/test_slider.py::test_min 
01-20 09:37:08.603  3201  3245 I python.stdout: SKIPPED (value is 0.0)            [ 92%]
01-20 09:37:08.603  3201  3245 I python.stdout: 
01-20 09:37:08.603  3201  3245 I python.stdout: tests/widgets/test_slider.py::test_max 
01-20 09:37:08.604  3201  3245 I python.stdout: SKIPPED (value is 0.0)            [100%]
01-20 09:37:08.609  3201  3245 I python.stdout: 
01-20 09:37:08.609  3201  3245 I python.stdout: 
01-20 09:37:08.609  3201  3245 I python.stdout: ==================================== PASSES ====================================
01-20 09:37:08.609  3201  3245 I python.stdout: ========================= 5 passed, 8 skipped in 0.51s =========================
```

That's definitely an improvement on the old behavior, but it's still a bit more "vertically verbose" than comparable output on the platforms.

Based on the behavior, I'm guessing that you're doing a write to log on `flush()`? Is there a particular reason for outputting on flush, rather than newline?

**Comment by mhsmith** [comment]
> Based on the behavior, I'm guessing that you're doing a write to log on `flush()`? Is there a particular reason for outputting on flush, rather than newline?

It now outputs on `flush` or newline, whichever comes first. The reasoning is that if the user calls `flush`, that means they want to see the output immediately, and the only way we can do that is to output a partial line.

**Comment by mhsmith** [comment]
Previously it would output on every call to `write`, which would break up lines much more.

File: {{ cookiecutter.safe_formal_name }}/app/build.gradle [hunk 1]
```diff
@@ -27,11 +24,10 @@ android {
...
             }
-        }
-        externalNativeBuild {
-            cmake {
-                cppFlags "-std=c++14"
-            }
+
+            // pytest and unittest both require individual .py files for test discovery.
+            // This is the package name used by briefcase-template.
+            extractPackages "tests"
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply {{ cookiecutter.safe_formal_name }}/app/build.gradle]: OK, I've changed it to use `test_sources`. This requires https://github.com/beeware/briefcase/pull/1061.

For any other use case, I think `build_gradle_extra_content` is fine for now.

**Comment by mhsmith** [comment]
With the updates to this PR, https://github.com/chaquo/chaquopy/issues/777, and other stuff, I didn't manage to complete the Chaquopy public release today. If nothing else comes up in final testing, it'll be available tomorrow.

**Comment by freakboy3742** [comment]
> It now outputs on `flush` or newline, whichever comes first. The reasoning is that if the user calls `flush`, that means they want to see the output immediately, and the only way we can do that is to output a partial line.
> 
> Previously it would output on every call to `write`, which would break up lines much more.

That's what I figured. I guess it's a bit of a value judgement - we have to make a judgement call between making the final output "clean", or capturing the immediacy for those small number of cases where it will matter (e.g., outputting the name of the test that will be run *before* it crashes). The `std-nslog` wrapper only honors newline at present; I guess we can use the comparison between the two as a validation of which approach works better in practice.

**Review by freakboy3742 — APPROVED** [review]

**Pull request merged**